packages/devtools_app/lib/src/standalone_ui/api/impl/dart_tooling_api.dart

```diff
@@ -82,43 +86,153 @@ class DartToolingApiImpl implements DartToolingApi {
   }
 }
 
-/// Base class for the different APIs that may be available.
-abstract base class ToolApiImpl {
-  ToolApiImpl(this.rpc);
-
-  static Future<Map<String, Object?>?> tryGetCapabilities(
-    json_rpc_2.Peer rpc,
-    String apiName,
-  ) async {
-    try {
-      final response = await rpc.sendRequest('$apiName.getCapabilities')
-          as Map<Object?, Object?>;
-      return response.cast<String, Object?>();
-    } catch (_) {
-      // Any error initializing should disable this functionality.
-      return null;
-    }
+/// A client for interacting with the legacy postMessage API with the same
+/// interface as the newer DTD-based [EditorClient].
+///
+/// This class allows the sidebar to use the new [EditorClient] APIs while still
+/// being compatible with postMessage based clients.
+class PostMessageEditorClient implements EditorClient {
+  PostMessageEditorClient(this._api) {
+    // In PostMessage world, we just get events with the entire new list so
+    // we must figure out what the actual changes are so we can produce the
+    // same kinds of events as the new DTD version.
+    _api.devicesChanged.listen((e) {
+      final supportedDevices = e.devices.map(
+        (d) => EditorDevice.fromJson({...d.toJson(), 'supported': true}),
+      );
+      final unsupportedDevices = e.unsupportedDevices?.map(
+            (d) => EditorDevice.fromJson({...d.toJson(), 'supported': false}),
+          ) ??
+          [];
+      final newDevices =
+          supportedDevices.followedBy(unsupportedDevices).toList();
+      final newIds = newDevices.map((d) => d.id).toSet();
+      final oldIds = _currentDevices.map((d) => d.id).toSet();
+
+      // Devices that are not in the new set have been removed.
+      for (final id in oldIds.difference(newIds)) {
+        _eventController.add(DeviceRemovedEvent(deviceId: id));
+      }
+      // Devices in the new set have either been changed or were added.
+      for (final device in newDevices) {
+        if (oldIds.contains(device.id)) {
+          _eventController.add(DeviceChangedEvent(device: device));
+        } else {
+          _eventController.add(DeviceAddedEvent(device: device));
+        }
+      }
+      // And record the updated set.
+      _currentDevices
+        ..clear()
+        ..addAll(newDevices);
+
+      // Finally, handle if the selection changed.
+      if (e.selectedDeviceId != _currentSelectedDeviceId) {
+        _currentSelectedDeviceId = e.selectedDeviceId;
+        _eventController.add(
+          DeviceSelectedEvent(deviceId: _currentSelectedDeviceId),
+        );
+      }
```

Comment:
nit: can you put this handler into a private method. Same for below.

Reply:
Done in 0f63b7d59